modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/actions/addFragment.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {fetch} from 'frontend-js-web';
+import {LAYOUT_DATA_ITEM_TYPES} from '../config/constants/layoutDataItemTypes';
+import {ADD_FRAGMENT_ENTRY_LINK_AND_ITEM} from './types';
+
+function addFragmentEntryLinkAndItem({
+	fragmentEntryLink,
+	itemId,
+	itemType,
+	parentId,
+	position
+}) {
+	const {
+		fragmentEntryLinkId
+	} = fragmentEntryLink;
+
+	return {
+		config: {
+			fragmentEntryLinkId,
+		},
+		fragmentEntryLinkId,
+		fragmentEntryLink,
+		itemId,
+		itemType,
+		parentId,
+		position,
+		type: ADD_FRAGMENT_ENTRY_LINK_AND_ITEM
+	};
+}
+
+export default function addFragment({
+	config,
+	store,
+	fragmentKey,
+	fragmentGroupId,
+	parentId,
+	position
+}) {
+	return (dispatch) => {
+		const {
+			portletNamespace
+		} = config;
+
+		const {
+			addFragmentEntryLinkURL,
+			classNameId,
+			classPK,
+			segmentsExperienceId
+		} = store;
+
+		const formData = new FormData();
+		formData.append(`${portletNamespace}fragmentKey`, fragmentKey);
+		formData.append(`${portletNamespace}classNameId`, classNameId);
+		formData.append(`${portletNamespace}classPK`, classPK);
+		formData.append(`${portletNamespace}groupId`, fragmentGroupId);
+		formData.append(`${portletNamespace}segmentsExperienceId`, segmentsExperienceId);
+
+		fetch(addFragmentEntryLinkURL, {
+			body: formData,
+			method: 'POST'
+		})
+			.then(response => response.json())
+			.then(fragmentEntryLink => {
```

Comment:
hey @julien I've been thinking a bit about this if we could do the application offline first, I know there are a lot of things that depend on the backend, but it could help to improve the experience. Offline first here does not mean that you would not need the backend but if the internet goes down the person still keeps editing the fragments and adding them.

For example, adding a `Button` fragment, if this fragment information already exists we can add it to the store without waiting for the backend.

Reply:
@matuzalemsteles as I've said before the communication with backend is not 100% clear, I guess we could take some time to analyse the previous version of the app but so far I've not been able to add an item without the backend - although I agree it would be nice.

@p2kmgcl what do you think?

Reply:
@julien, yeah i was thinking better about it, i think we have a big backend dependency to solve the stuff so it's probably not possible to do anything in case we add the sidebar fragment to the page. We can think about it later when we have with all the implemented behaviors maybe.